### PR TITLE
Review report

### PR DESCRIPTION
**Pull request by aanand-1706**
### 🔴 Critical Violations

**Missing Test Coverage (new production code, no tests)**
- Substantial new code (ReportsController + Reports::* classes and new ActiveRecord scopes) added with no corresponding specs. Repository search found no specs referencing ReportsController, Reports::BaseReport, Reports::ReviewReport, or the new scopes.
  - Required specs to add:
    - spec/controllers/reports_controller_spec.rb — auth, valid/invalid type, 404/422/500 branches.
    - spec/models/reports/base_report_spec.rb — streaming/reducer contract (source/initial_state/accumulate/finalize), run semantics.
    - spec/models/reports/review_report_spec.rb — ReviewersReducer, ScoresReducer, AvgRangesReducer behavior and integration.
    - Specs for Response.submitted_review_responses_for and ReviewResponseMap.for_assignment.

**Authorization & Error-Path Tests Missing**
- ReportsController.action_allowed? restricts to assignment TAs but there are no authorization tests to verify access control (TA vs non-TA).
- No tests for error branches: missing assignment (404), invalid report type (422), and generic rescue (500).

### 🟡 Medium Severity Issues

**Overly Broad Error Handling**
- ReportsController rescues StandardError and returns 500 JSON. This masks unexpected failures and reduces observability. Prefer rescuing specific exceptions or letting unexpected errors surface (with logging/monitoring) and test error behavior.

**Unrelated/Whitespace Churn**
- Raw changes included a whitespace-only edit in app/models/Item.rb (move to separate PR to avoid noise). Routes file shows formatting changes plus a new POST /reports/fetch_report (config/routes.rb:199) — minimize non-functional formatting churn.

**Potential Performance Concerns**
- New streaming reducer pattern is reasonable, but no performance or query-count tests. Ensure eager-loading covers all associations used by reducers and add tests/assertions for N+1 or memory growth on larger fixtures.

**Danger/Debrief Signals Present**
- Repo contains TODO comments (responses_controller, authorization) and existing debug patterns that the Danger rules check for. This PR modifies app/ without adding specs, which will trigger the repository's Dangerfile checks.

### 🟢 Low Severity / Checklist

- No schema or migration files were added; confirm referenced DB columns/indexes (reviewed_object_id, response_map_id, round, any aggregate fields) already exist — db/schema.rb unchanged.
- New routes added: POST /reports/fetch_report present in config/routes.rb.
- No debug prints introduced in the new files per current search.

### Summary / Recommendation

Status: CANNOT MERGE

Required before merge:
1. Add comprehensive specs (controller + model reducers + scope tests) listed above.
2. Add authorization tests for TA access and negative cases.
3. Replace the broad StandardError rescue with narrower handling or re-raise after logging; add tests for error branches.
4. Remove or separate unrelated whitespace change (Item.rb) into its own PR and revert unnecessary formatting churn in routes.
5. Add basic performance/query-count tests or benchmarks for reducer flows to guard against N+1/memory issues.

The implementation and reducer design are plausible, but Expertiza Danger policy and repo standards require test coverage, tighter error handling, and avoidance of unrelated churn before approval.

[Review Change Stack](https://app.coderabbit.ai/change-stack/expertiza/reimplementation-back-end/pull/351?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

### CODE REVIEW

File: app/controllers/reports_controller.rb [hunk 1]
```diff
@@ -0,0 +1,39 @@
+# frozen_string_literal: true
+
+class ReportsController < ApplicationController
+  REPORT_CLASSES = {
+    'review_response_map'           => Reports::ReviewReport,
+    'feedback_response_map'         => Reports::FeedbackReport,
+    'teammate_review_response_map'  => Reports::TeammateReviewReport,
+    'bookmark_rating_response_map'  => Reports::BookmarkRatingReport,
+    'answer_tagging'                => Reports::AnswerTaggingReport,
+    'basic'                         => Reports::BasicReport
```
Comment (efg) [inline app/controllers/reports_controller.rb]: Why shouldn't 'basic' be first in the list?

File: app/controllers/reports_controller.rb [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+  }.freeze
+
+  # Only TAs, instructors, and admins may view reports.
+  def action_allowed?
+    current_user_has_ta_privileges?
```
Comment (efg) [inline app/controllers/reports_controller.rb]: This seems to be wrong.  It allows any instructor and any TA to see any report, if they know the URL.

File: app/controllers/reports_controller.rb [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+  end
+
+  # POST /reports/fetch_response_report
+  # Returns the requested report as JSON.
+  def fetch_response_report
```
Comment (efg) [inline app/controllers/reports_controller.rb]: Why is the method name "fetch_response_report" when not all of the reports are called response reports?

File: app/controllers/reports_controller.rb [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+    assignment_id = params[:assignment_id]
+    type = params[:type] || 'basic'
+
+    report_class = REPORT_CLASSES[type]
+    unless report_class
```
Comment (efg) [inline app/controllers/reports_controller.rb:23]: I question whether this clause is useful.  It seems to be testing whether the app is working correctly.  That kind of check should be done by automated tests, not application code within the app.

File: app/controllers/reports_controller.rb [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+      return render json: {
+        error: "Unknown report type: #{type}. Valid types: #{REPORT_CLASSES.keys.join(', ')}"
+      }, status: :unprocessable_entity
+    end
+
+    assignment = Assignment.find(assignment_id)
+    data = report_class.for_assignment(assignment).run
+    render json: { type: type, assignment_id: assignment_id.to_i }.merge(data)
+  rescue ActiveRecord::RecordNotFound
+    render json: { error: 'Assignment not found' }, status: :not_found
```
Comment (efg) [inline app/controllers/reports_controller.rb]: Similar question here.  Is there any way someone could invoke an operation from the UI that would fail to find an assignment?

File: app/models/reports/base_reducer.rb [hunk 6]
```diff
@@ -0,0 +1,78 @@
+# frozen_string_literal: true
+
+module Reports
+  # Template for a streaming reduce-based report.
+  #
+  # Design rationale (addresses two anti-patterns from the naive approach):
```
Comment (efg) [inline app/models/reports/base_reducer.rb:6]: I think you mean, "reduction-based report".

File: app/models/reports/review_report.rb [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+  # ReviewersReducer and ScoresReducer share a single state keyed by reviewer_id
+  # so reviewer info and scores are co-located without a merge loop:
+  #   { reviewer_id => { id:, user_id:, name:, handle:,
+  #                      scores: { reviewee_id => { round => score_pct } } } }
+  #
+  # AvgRangesReducer runs independently:
+  #   { team_id => avg_score }
+  #
+  # Each reducer streams its source via find_each — no full result set is ever
+  # materialised in Ruby at once.
```
Comment (efg) [inline app/models/reports/review_report.rb]: Please use American spelling in comments.

File: app/models/reports/review_report.rb [hunk 8]
```diff
@@ -0,0 +1,121 @@
+# frozen_string_literal: true
+
+module Reports
```
Comment (efg) [inline app/models/reports/review_report.rb:3]: I'm confused about why there are two modules called Reports, in different classes.  As the Ruby program is executed, the second definition will replace the first definition.

Also, if the Reports module is meant to be used by multiple reports, it seems that it should be in base_report.rb.

File: app/models/reports/review_report.rb [hunk 9]
```diff
@@ -0,0 +1,121 @@
...
+  # Peer-review report composed of three streaming reducers.
+  #
+  # ReviewersReducer and ScoresReducer share a single state keyed by reviewer_id
```
Comment (efg) [inline app/models/reports/review_report.rb]: I think it would be helpful in the comments to say what the ReviewersReducer does and what the ScoresReducer does.  The comment may imply that, but it would be better to state explicitly.

File: app/models/reports/review_report.rb [hunk 10]
```diff
@@ -0,0 +1,121 @@
...
+  class ReviewReport
+    def self.for_assignment(assignment)
+      new(assignment)
+    end
+
+    def initialize(reportable)
+      @reportable = reportable
+    end
+
+    def run
```
Comment (efg) [inline app/models/reports/review_report.rb:96]: Needs a method comment.

File: app/models/reports/review_report.rb [hunk 11]
```diff
@@ -0,0 +1,121 @@
...
+
+        state[reviewer_id][:scores][reviewee_id][round] = score_pct
+      end
+    end
+
+    # -----------------------------------------------------------------------
+    # Reducer 3 — per-team average review score.
+    # Streams AssignmentTeam rows with review_mappings, responses, and scores
+    # eagerly loaded to avoid N+1. Delegates score computation to
+    # aggregate_review_grade (via ReviewAggregator concern) which picks the
```
Comment (efg) [inline app/models/reports/review_report.rb]: Should that be "ReviewReducer"?

File: app/models/response.rb [hunk 12]
```diff
@@ -5,6 +5,20 @@ class Response < ApplicationRecord
   include MetricHelper
 
   belongs_to :response_map, class_name: 'ResponseMap', foreign_key: 'map_id', inverse_of: false
+
+  scope :submitted_review_responses_for, ->(assignment_id) {
+    # Subquery picks the latest submitted response per (response_map, round)
```
Comment (efg) [inline app/models/response.rb]: I think there should be a method (e.g., "latest_response") that returns the latest response for a round.  That would make the code clearer.

File: config/routes.rb [hunk 13]
```diff
@@ -77,37 +77,37 @@
...
-          delete :delete_all_for_reviewer
-        end
-
-        member do
-          patch :submit_review
-          patch :unsubmit_review
-          patch :grade_review
-          delete :delete_mapping
-        end
+     resources :review_mappings, only: [] do
```
Comment (efg) [inline config/routes.rb]: Most of the changes in this file are whitespace only.  Removing those changes, or making them in a separate commit, would make the diff easier to follow.

File: config/routes.rb [hunk 14]
```diff
@@ -219,4 +225,4 @@
       resources :assignments do
         resources :duties, controller: 'assignments_duties', only: [:index, :create, :destroy]
       end
-end
+end
```
Comment (efg) [inline config/routes.rb]: Please add an EoL.

**Review by efg — COMMENTED** [review]
Good job!  Please see if my suggestions help improve the code.

File: app/controllers/reports_controller.rb (hunk 1 above)
Comment (aanand-1706) [reply app/controllers/reports_controller.rb]: Moved to the top

File: app/controllers/reports_controller.rb (hunk 2 above)
Comment (aanand-1706) [reply app/controllers/reports_controller.rb]: Updated to have `current_user_teaching_staff_of_assignment?`

File: app/controllers/reports_controller.rb (hunk 3 above)
Comment (aanand-1706) [reply app/controllers/reports_controller.rb]: I was referring to all reports being related to responses submitted by a user. Changed to `fetch_report`

File: app/controllers/reports_controller.rb (hunk 4 above)
Comment (aanand-1706) [reply app/controllers/reports_controller.rb:23]: The check isn't testing internal app correctness - type comes from user-supplied POST params, which means an unknown value is a realistic scenario. Without the guard, passing an invalid type would hit `nil.for_assignment` and raise a `NoMethodError`.

File: app/controllers/reports_controller.rb (hunk 5 above)
Comment (aanand-1706) [reply app/controllers/reports_controller.rb]: In the ideal scenario, a person shouldn't be able to invoke an operation with an invalid assignment. But since the APIs could be called directly (i.e. curl, Postman, etc.), BE should validate independently.

File: app/models/reports/base_reducer.rb (hunk 6 above)
Comment (aanand-1706) [reply app/models/reports/base_reducer.rb:6]: Updated

File: app/models/reports/review_report.rb (hunk 8 above)
Comment (aanand-1706) [reply app/models/reports/review_report.rb:3]: When `module Reports` appears in multiple files, all definitions are merged into the same `Reports` namespace. `Reports::BaseReport`, `Reports::ReviewReport`, etc. all coexist in the same module. 

`module Reports` in multiple files reopens and extends the same namespace rather than replacing it.

File: app/models/reports/review_report.rb (hunk 9 above)
Comment (aanand-1706) [reply app/models/reports/review_report.rb]: Updated the comments

File: app/models/reports/review_report.rb (hunk 7 above)
Comment (aanand-1706) [reply app/models/reports/review_report.rb]: Updated

File: app/models/reports/review_report.rb (hunk 10 above)
Comment (aanand-1706) [reply app/models/reports/review_report.rb:96]: Updated

File: app/models/reports/review_report.rb (hunk 11 above)
Comment (aanand-1706) [reply app/models/reports/review_report.rb]: This is an already existing module being used in several other places as well. Should I update the module name and corresponding usages?

File: app/models/response.rb (hunk 12 above)
Comment (aanand-1706) [reply app/models/response.rb]: The `submitted_review_responses_for` scope already encapsulates this logic at the SQL level using a `MAX(id)` subquery grouped by `(response_map_id, round)`, which fetches the latest submitted       
  response per round for all response maps in a single query. Extracting a separate `latest_response` method would either duplicate this logic or require calling it per response map in a loop,       
  reintroducing N+1 queries.

File: app/models/reports/review_report.rb (hunk 8 above)
Comment (aanand-1706) [reply app/models/reports/review_report.rb:3]: Tried removing modules wrapper for the review report, but was getting errors related to uninitialized constant. It's because we have kept all report related files in a directory called reports. 

This was the issue: 
```
Zeitwerk — Rails' autoloader — maps file paths to constant names:                                                                                                                                  
                                                                                                                                                                                                   
  app/models/reports/review_report.rb  →  Reports::ReviewReport                                                                                                                                      
  app/models/reports/base_reducer.rb   →  Reports::BaseReducer                                                                                                                                       
                                                                                                                                                                                                     
  If you remove module Reports and just write class ReviewReport, Zeitwerk still 
expects the constant Reports::ReviewReport to be defined in that file (because of the path). 
It won't find a bare   ReviewReport there — it would look in app/models/review_report.rb for that. 
So you'd get uninitialized constant Reports::ReviewReport at boot.                                                   
                                                                                                                                                                                                     
  The module Reports wrapper is not a design choice here - it's a requirement imposed by the file location.
The name of the enclosing directory is the module name. 
If you genuinely don't want the  
  module, you'd have to move all the report files up to app/models/ directly, 
but then they'd all be in the global namespace alongside Assignment, User, etc.
                                                                                                                                                                                                     
  Every subdirectory in this codebase follows the same rule - 
app/models/strategies/, app/models/review_mapping_strategies/, app/models/analytic/ 
all have matching module wrappers for the same reason.
```

If we don't want the modules wrapper, we will have to keep all report files at models level.
